src/query/functions/aggregation/take.go

```diff
@@ -147,34 +162,95 @@ func (n *takeNode) ProcessBlock(queryCtx *models.QueryContext, ID parser.NodeID,
 	}
 
 	for index := 0; stepIter.Next(); index++ {
-		step := stepIter.Current()
-		values := step.Values()
-		aggregatedValues := n.op.takeFunc(values, buckets)
-		if err := builder.AppendValues(index, aggregatedValues); err != nil {
+		values := stepIter.Current().Values()
+		if err := builder.AppendValues(index, n.op.takeFunc(heap, values, buckets)); err != nil {
 			return nil, err
 		}
 	}
-
 	if err = stepIter.Err(); err != nil {
 		return nil, err
 	}
-
 	return builder.Build(), nil
 }
 
-// shortcut to return empty when taking <= 0 values
-func takeNone(values []float64, buckets [][]int) []float64 {
-	util.Memset(values, math.NaN())
-	return values
+func maxSeriesCount(buckets [][]int) int {
+	result := 0
+
+	for _, bucket := range buckets {
+		if len(bucket) > result {
+			result = len(bucket)
+		}
+	}
+
+	return result
+}
+
+func (n *takeNode) processBlockInstantaneous(
+	heap utils.FloatHeap,
+	queryCtx *models.QueryContext,
+	metadata block.Metadata,
+	stepIter block.StepIter,
+	seriesMetas []block.SeriesMeta,
+	buckets [][]int) (block.Block, error) {
+	ixLastStep := stepIter.StepCount() - 1 //we only care for the last step values for the instant query
+	for i := 0; i <= ixLastStep; i++ {
+		if !stepIter.Next() {
+			return nil, fmt.Errorf("invalid step count; expected %d got %d", stepIter.StepCount(), i+1)
+		}
+	}
+	metadata.ResultMetadata.KeepNaNs = true
+	values := stepIter.Current().Values()
+	takenSortedValues := n.op.takeInstantFunc(heap, values, buckets, seriesMetas)
+	blockValues, blockSeries := mapToValuesAndSeriesMetas(takenSortedValues)
+
+	//adjust bounds to contain single step
+	time, err := metadata.Bounds.TimeForIndex(ixLastStep)
+	if err != nil {
+		return nil, err
+	}
+	metadata.Bounds = models.Bounds{
+		Start:    time,
+		Duration: metadata.Bounds.StepSize,
+		StepSize: metadata.Bounds.StepSize,
+	}
+
+	blockBuilder, err := n.controller.BlockBuilder(queryCtx, metadata, blockSeries)
+	if err != nil {
+		return nil, err
+	}
+	if err = blockBuilder.AddCols(1); err != nil {
+		return nil, err
+	}
+	if err := blockBuilder.AppendValues(0, blockValues); err != nil {
+		return nil, err
+	}
+	if err = stepIter.Err(); err != nil {
+		return nil, err
+	}
+	return blockBuilder.Build(), nil
+}
+
+func mapToValuesAndSeriesMetas(takenSortedValues []valueAndMeta) ([]float64, []block.SeriesMeta) {
+	blockValues := make([]float64, len(takenSortedValues))
+	blockSeries := make([]block.SeriesMeta, len(takenSortedValues))
+	for i, sortedValue := range takenSortedValues {
+		blockValues[i] = sortedValue.val
+		blockSeries[i] = sortedValue.seriesMeta
+	}
```

Comment:
Safer to use append here

Reply:
+1

Reply:
Here we do not care about indices so changed to `append`.